File: labs/11/clojure/src/threadedArraySum.clj
;;;; threadedarraysum.clj sums the integers in a file using an array.
;;;;
;;;; Usage: clojure -m threadedarraysum <inputFileName> 
;;;;
;;;; Input: the name of the input file and the number of threads
;;;;         (both entered on the command-line).
;;;; Precondition: the first line of the file is an integer N
;;;;               the second line is a blank
;;;;               the remaining lines contain N integers
;;;;               (one per line).
;;;; Output: the sum of the other numbers in the file
;;;;          and how long it took to compute that sum.
;;;;
;;;; Begun by: Prof. Adams for CS 214 at Calvin College.
;;;; Completed by: Jacob Brink
;;;; Date: 5/4/2020
;;;; Lab: 11
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns threadedArraySum
  (:require [clojure.core.async            ; need this clause to use async
             :refer [chan go put! take! <! >! <!! >!!] 
             ] 
            ) 
  )

;;; recursively fill an array with values from a Java Scanner
;;;
;;; Parameters: array, an array of longs 
;;;             in, a scanner.
;;; Precondition: in contains N values &&
;;;               array contains space for N values.
;;; Postcondition: array contains the N values from in.
;;; Pass back: array, containing the N values from in.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn fill [array in] 
  (loop [i 0]                            ; set i to 0
    (when (< i (count array))            ; while i < array.size
      (aset array i (.nextLong in))      ;  array[i] = in.nextLong
      (recur (inc i))                    ;  recurse (i+1)
      )
    )
  )

;; The following version of fill works for smaller arrays,
;;  but generates StackOverflow errors for larger arrays
;;  even though the function uses end-recursion.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
                                        ;(defn fill [ array index in] 
                                        ;  (if (< index (count array))
                                        ;    (do
                                        ;        (aset-long array index (.nextLong in))
                                        ;        (fill array (inc index) in)
                                        ;    )
                                        ;    array
                                        ;  )
                                        ;)


;;; read values from a file into an array 
;;;
;;;   Receive: inFile, a String naming an input file.
;;;   Precondition: the first line of the input file is N,
;;;                  the second line is blank, and
;;;                  the rest of the file contains N integers,
;;;                   one per line.
;;;   Postcondition: the N int values have been read
;;;                   from 'in' and stored in an array.
;;;   Return: an array containing the N values from the file.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn readFile [fileName]
                                        ; before proceeding, check to make sure fileName exists 
  (assert (.exists (clojure.java.io/file fileName))
          (str "input file '" fileName "' not found"))
  (let
      [
       in (java.util.Scanner. (java.io.File. fileName))
       numNumbers (.nextLong in)       ; read first line
       blankLine (.nextLine in)        ; read blank line
       array (long-array numNumbers 0) ; allocate array of N zeros
       ]

    (fill array in)                  ; fill array with other vals from in
    (.close in)                      ; close Scanner
    array                            ; return the array
    )
  )





;;; calcPartialSum():
;;; - sum a 'slice' of an array given start and stop indices.
;;; 
;;; Receive: anArray, an array containing values to be summed;
;;;          start, a long containing the starting index;
;;;          stop, a long containing the stopping index.
;;; Return: the sum of anArray[start] .. anArray[stop-1] (inclusive).
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn calcPartialSum [anArray start stop]
(loop
    [                           ; initially:
      partialSum 0              ;  partialSum = 0
      i          start          ;  i = start
    ]
    (if (>= i stop)             ; Basis: i >= stop:
      partialSum                ;  return partialSum
                                ; I-step:
                                ;  return loop(partialSum + a[i], ++i)
      (recur (+ partialSum (get anArray i)) (inc i))
    )
  )
)

;;; sum a 'slice' of an array based on thread ID and numThreads
;;;
;;; Receive: anArray, the array of ints to be summed;
;;;          id, the ID of the current thread;
;;;          numThreads, how many workers we have;
;;;          channel, the channel to write its slice-sum to.
;;; Postcondition: the sum of the values in this thread's slice
;;;           of anArray have been put into channel.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; 
(defn sumSlice [anArray id numThreads channel]
  (let
    [ arraySize   (count anArray)              ; determine array size
      sliceSize   (quot arraySize numThreads)  ; calculate slice size
      start       (* id sliceSize)             ; calc. starting index
      stop        (if (< id (- numThreads 1))  ; calc. stopping index
                     (+ start sliceSize)       ; - all but last thread
                     arraySize                 ; - last thread
                  )
                                               ; calc. sum of my slice
      partialSum  (calcPartialSum anArray start stop)
    ]

    ; write partial sum to channel
    (put! channel partialSum)
  )
  )



;;; sum the values in an array sequentially.
;;; 
;;; Receive: anArray, an array containing values to be summed.
;;; Return: the sum of the values in anArray.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn sumArray [anArray  numThreads]
  (let
      [ resChannel (chan) ]        ; define shared channel for results

    (loop [id 1]                 ; loop to fork threads,
      (when (< id numThreads)    ;  each writing its partial-sum 
        (go                      ;  to the channel
          (sumSlice anArray id numThreads resChannel)
          )
        (recur (inc id))
        )
      )
                                        ; main thread does slice 0
    (sumSlice anArray 0 numThreads resChannel)
                                        ; loop to read each partial-sum
    (loop [ sum 0                ;  from resChannel and add it to sum
           id  0]               ;  
      (if (>= id numThreads)     ; Basis: id >= numThreads:
        sum                      ;  return sum
                                        ; I-Step: 
                                        ;  return sum + next resChannel val
                                        ;          and recurse (id+1)
        (recur (+ sum (<!! resChannel)) (inc id)) 
        )
      )
    )
  )

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; main method 
;;;
;;; Receive: inFile, a String containing the name of an input file.
;;; Precondition: inFile has N+2 lines, formatted as follows:
;;;                the first line is a positive integer N 
;;;                the second line is a blank line
;;;                the remaining N lines are the values (1 per line).
;;; Output: the sum of the values in inFile
;;;           and the time required to sum them
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main [inFile numThreadsStr]
  (let
      [
       anArray    (readFile inFile)        ; read inFile into anArray 
       numValues  (count anArray)          ; determine numValues
       numThreads   (read-string numThreadsStr)  ; numThreadsStr -> integer
       
       startTime  (System/nanoTime)        ; - start timer
       sum        (sumArray anArray numThreads)       ; - sum values sequentially
       stopTime   (System/nanoTime)        ; - stop timer
       totalTime  (- stopTime startTime)   ; - calc. sequential time
       ]
                                        ; output results
    (printf "\nThe *parallel* sum of the %d numbers is %d;\n"
            numValues sum)
    (printf " summing them with %d thread(s) took %d time units.\n\n"
            numThreads totalTime)
    )
  )

File: labs/11/java/ThreadedArraySum.java
/* ThreadedArraySum.java sums the integers in a file using an array.
 *
 * Usage: java -Xint ThreadedArraySum <fileName> <threadCount>
 *
 * Input: the name of the input file
 *         (entered on the command-line).
 * Precondition: the first line of the file is the number
 *                of (other) numbers in the file.
 * Output: the sum of the other numbers in the file
 *          and how long it took to compute that sum.
 *
 * Begun by: Prof. Adams for CS 214 at Calvin College.
 * Completed by: Jacob Brink
 * Date: 5/4/2020
 * Lab: 11
 */

import java.util.Scanner;                   // convenient I/O
import java.io.File;                        // file I/O
import java.io.FileNotFoundException;       // file exception
import java.lang.System;                    // nanoTime()

public class ThreadedArraySum {

    private class Helper extends Thread {

	public Helper(int id) {
	    super();
	    myID = id;
	    myPartialSum = 0;
	}

	public void run() {
	    myPartialSum = sumSlice(myID);
	}

	public final long getPartialSum() {
	    return myPartialSum;
	}

	private int  myID = 0;
	private long myPartialSum = 0;
    } // Helper
    
    private int myNumThreads;
    
    public static void main(String [] args) {
	ThreadedArraySum me = new ThreadedArraySum(args);
	me.run();
    }

    /* constructor
     * @param: args, an array of Strings
     * Precondition: args.length == 1 &&
     *                args[0] is the name of an input file
     *                 whose first line is N, an integer
     *                 followed by a blank line
     *                 followed by N integers, one per line.
     * Postcondition: myScanner is a Scanner for that input file
     *             && myArray is an array containing the N ints.
     */
    public ThreadedArraySum(String [] args) {
	if (args.length < 1) {
	    System.err.println("\n*** Usage: java Sum <fileName> <threadCount>\n");
	    System.exit(1);
	}
	Scanner aScanner = openScanner( args[0] ); // open input file
	myArray = readFile(aScanner);              // fill array with values
	myNumThreads = getNumThreads( args );
    }

    private long sumSlice(int id) {
        int sliceSize = myArray.length / myNumThreads;
        int start = id * sliceSize;         // starting index
        int stop = (id+1) * sliceSize;      // stopping index
        if ( id == myNumThreads-1 ) {       // have final thread
            stop = myArray.length;            //  handle leftovers
        } 
        long sliceSum = 0;
        for (int i = start; i < stop; ++i) {  // sum the ints
            sliceSum += myArray[i];           //  in my slice

        }
        return sliceSum;
    }

    private final int getNumThreads( String [] args ) {
	if (args.length >= 2) {
	    return Integer.parseInt( args[1] );
	} else {
	    return 1;
	}
    }
    
    /* utility to open / return a Scanner to a file 
     *   whose name comes from the command-line.
     * 
     * @param: fileName, a String
     * Precondition: fileName is the name of an input file.
     * Return: a Scanner for that input file.
     */
    private Scanner openScanner(String fileName) {
	Scanner in = null;
	try {
	    File theFile = new File ( fileName );
	    in = new Scanner( theFile );
	} catch ( FileNotFoundException  fnfe ) {
	    System.err.println("\n*** Unable to open '" + fileName + "'\n");
	    System.exit(1);
	}

	return in;
    }

    /* fill an array with values from a Scanner
     * @param: in, a Scanner to an input file.
     * Precondition: the first line of the input file is N,
     *                the number of (other) numbers in the file.
     * Postcondition: the N int values have been read
     *                 from 'in' and stored in an array.
     * Return: the array containing the N int values.
     */
    private int[] readFile(Scanner in) {
	int numNumbers = in.nextInt();           // read how many numbers
	int [] array = new int[numNumbers];      // allocate array
	for (int i = 0; i < array.length; i++) { // fill array
	    array[i] = in.nextInt();              //   with numbers
	}
	in.close();                              // close input file
	return array;                            // we're done
    }

    /* how long does it take to process the values in myArray?
     * Precondition: myArray has been filled with N input values.
     * Output: the sum of the values in myArray
     *          and how long it took to compute that sum.
     */
    public void run() {
	long startTime = System.nanoTime();             // start timing
	long sum = sumArray(myArray);                    // sum the values
	long totalTime = System.nanoTime() - startTime; // stop timing
     
	System.out.println("\nThe sum of the " + myArray.length
			   + " numbers is: " + sum 
			   + "\n and summing them took " 
			   + totalTime + " time units.\n");
    }

    /* sum the values in an array
     * @param: arr, an array of integers
     * Return: the sum of the int values in arr.
     */
    private long sumArray(int[] arr) {
	Helper[] helpers = new Helper[myNumThreads];

	for (int i = 1; i < myNumThreads; ++i) {     // for each helper:
	    helpers[i] = new Helper(i);                  //  create, and
	    helpers[i].start();                          //  launch them
	}

	long sum = sumSlice(0);                         // main thread does slice 0

	try {
	    for (int i = 1; i < myNumThreads; ++i) {  // for each helper h:
		helpers[i].join();
		sum += helpers[i].getPartialSum();        //  get its partial sum
	    }
	} catch( InterruptedException ie) {             // required by join()
	    System.err.println("\n*** a Helper was interrupted!\n");
	    System.err.println(ie);
	    System.exit(1);
	}

	return sum;
    }

    private int [] myArray = null;
}

